feat(rum): Add Salesforce LWC docs (#3058)
DataDog/integrations-extras: rum_salesforce_lwc/README.md

Reviewer: The “1. Add the static resource” and “2. Add the CSP Trusted Site” steps are identical across all setups.

Also, the Lightning Apps and Experience Cloud Head Markup sections differ by only one step, while the rest of the instructions are duplicated.

Could you explore a different layout that factors out the common steps

Author: I see your point but, being this a md, I opted to not make the user navigate back and forth and just follow a section to the end.  I'll see what I can do.

```

Once you have these, follow one of the deployment paths below:

- **[Lightning Apps](https://github.com/DataDog/browser-sdk/tree/main/packages/browser-rum-slim/src/salesforce#lightning-apps)**: Use a custom Lightning Web Component loaded from the Utility Bar.
- **[Experience Cloud Head Markup](https://github.com/DataDog/browser-sdk/tree/main/packages/browser-rum-slim/src/salesforce#experience-cloud-head-markup)**: Add the Datadog RUM initialization script directly to Head Markup.
- **[Experience Cloud Theme/Layout LWC](https://github.com/DataDog/browser-sdk/tree/main/packages/browser-rum-slim/src/salesforce#experience-cloud-themelayout-lwc)**: Add a custom Lightning Web Component to an Experience Builder page, shared region, or theme/layout area.

## Lightning Apps

Use this path for Salesforce LWC applications.

This path uses:

- A Salesforce Static Resource for the Datadog Salesforce RUM bundle.
- A CSP Trusted Site for the Datadog browser intake endpoint.
- A custom Lightning Web Component.
- A Utility Bar configuration with eager loading enabled.

### 1. Add the static resource

> Copy the Datadog Salesforce RUM bundle into your Salesforce project and register it as a static resource.

Reviewer: Where should the customer look to download it?

This path uses:

- A Salesforce Static Resource for the Datadog Salesforce RUM bundle.
- Experience Cloud CSP configuration.
- A Head Markup script that loads the SDK and tracks route changes.

### 1. Add the static resource

Copy the Datadog Salesforce RUM bundle into your Salesforce project and register it as a static resource.

Suggested configuration:

- Static Resource Name: `datadog_rum`
- File location: `staticresources/datadog_rum.js`
- Metadata location: `staticresources/datadog_rum.resource-meta.xml`

XML configuration:

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<StaticResource xmlns="http://soap.sforce.com/2006/04/metadata">
  <cacheControl>Public</cacheControl>
  <contentType>application/javascript</contentType>
</StaticResource>
```

If you configure this through the Salesforce UI:

1. Go to **Setup > Static Resources**.
2. Click **New**.
3. Set **Name** to `datadog_rum`.
4. Upload the Salesforce RUM JavaScript bundle.
5. Set **Cache Control** to `Public`.
6. Save the static resource.

> ### 2. Open the Experience Cloud site in Builder

Reviewer: <!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏I don't really understand this step. It seems to only indicate where to navigate?

# Salesforce Integration

## Overview

Instrument Salesforce Lightning Apps and Experience Cloud sites with Datadog Real User Monitoring using the Datadog Browser SDK slim RUM bundle.

> This guide covers four different deployment paths:

Reviewer 2: ```suggestion
This guide covers three deployment paths, plus a feature support reference:
```

- **[Lightning Apps](#lightning-apps)**: Use a custom Lightning Web Component loaded from the Utility Bar.
- **[Experience Cloud Head Markup](#experience-cloud-head-markup)**: Add the Datadog RUM initialization script directly to Head Markup.
- **[Experience Cloud Theme/Layout LWC](#experience-cloud-themelayout-lwc)**: Add a custom Lightning Web Component to an Experience Builder page, shared region, or theme/layout area.
- **[Feature Support](#salesforce-feature-support-matrix)**: Review supported features for the Datadog integration.

**Note**: Use only one deployment path per Salesforce app or Experience Cloud site.

## Setup

You will need the following Datadog RUM values:

- A Datadog RUM Application ID
- A Datadog RUM Client Token
- A Datadog site, such as `datadoghq.com`

> You can get those values in Datadog under **RUM > Digital Experience > Real User Monitoring > Manage Applications > Set Up Manually**.

Reviewer 2: ```suggestion
You can get those values in Datadog under **Digital Experience > Real User Monitoring > Manage Applications > Set Up Manually**.
```

Register the `datadogInit` component in your app's Utility Bar. You can do this through the Salesforce UI or by updating the app's FlexiPage metadata.

**Important**: set the `eager` property to `true`. Without this setting, the SDK initializes only when the user manually opens the Utility Bar item, which can cause lost initial page-load data.

File location: `flexipages/MyApp_UtilityBar.flexipage-meta.xml`

```xml
<componentInstance>
  <componentInstanceProperties>
    <name>eager</name>
    <type>decorator</type>
    <value>true</value>
  </componentInstanceProperties>
  <componentName>datadogInit</componentName>
  <identifier>datadogInit</identifier>
</componentInstance>
```

### 7. Validate the Lightning app installation

After deploying the component:

1. Open the Lightning app.
2. Open browser developer tools.
3. Confirm that the Datadog static resource loads successfully.
4. Confirm there are no CSP errors for the Datadog browser intake endpoint.
5. Navigate between Lightning pages.
6. In Datadog RUM Explorer, filter by the configured service and env.
7. Confirm that view events are created when navigation occurs.

## Experience Cloud Head Markup

> Use this path for Experience Cloud sites where you can add custom JavaScript through Head Markup. This is usually the simplest Experience Cloud approach because the SDK loads at the page level and does not rely on the Salesforce Utility Bar.

Reviewer 2: ```suggestion
Use this path for Experience Cloud sites where you can add custom JavaScript through Head Markup. For This is usually the most direct Experience Cloud approach because the SDK loads at the page level and does not rely on the Salesforce Utility Bar.
```

      trackLongTasks: true,
      trackResources: true,
      trackUserInteractions: true,
    })
  })
</script>
```

### 5. Publish the Experience Cloud site

After adding the Head Markup script:

1. Save the Head Markup configuration.
2. Publish the Experience Cloud site.
3. Open the published site in a new browser session.
4. Confirm that the Datadog static resource loads.
5. Confirm that no CSP errors appear in the browser console.
6. Navigate between site pages.
7. Confirm that RUM view events appear in Datadog.

## Experience Cloud Theme/Layout LWC

Use this path when you want to instrument an Experience Cloud site with a Lightning Web Component instead of Head Markup.

This path is useful when:

- Head Markup is unavailable.
- You want RUM configuration to be managed through Experience Builder component properties.
- You want to place the initializer in a shared region, template, page layout, or theme/layout area.
- You want to reuse the LWC-based implementation pattern across Salesforce surfaces.

Unlike Lightning Apps, Experience Cloud does not use the Utility Bar. The `datadogInit` component must be placed somewhere that loads on every page where RUM should run:

> - For simple installations, expose the component as an Experience Builder page component and place it in a shared region that appears across the site.

Reviewer 2: ```suggestion
- For lightweight installations, expose the component as an Experience Builder page component and place it in a shared region that appears across the site.
```

- Static Resource Name: `datadog_rum_slim`
- File location: `staticresources/datadog_rum_slim.js`
- Metadata location: `staticresources/datadog_rum_slim.resource-meta.xml`

XML configuration:

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<StaticResource xmlns="http://soap.sforce.com/2006/04/metadata">
  <cacheControl>Public</cacheControl>
  <contentType>application/javascript</contentType>
</StaticResource>
```

If you configure this through the Salesforce UI:

1. Go to **Setup > Static Resources**.
2. Click **New**.
3. Set **Name** to `datadog_rum_slim`.
4. Upload the slim RUM JavaScript bundle.
5. Set **Cache Control** to `Public`.
6. Save the static resource.

### 2. Configure CSP for the Datadog intake endpoint

In Experience Builder:

1. Go to **Settings > Security & Privacy**.
2. Change the security level from **Strict CSP** to **Relaxed CSP**.
3. Add the Datadog browser intake endpoint as a trusted site.

Use the following trusted site configuration for US1:

| Field | Value                              |
| ----- | ----------------------------------- |
| Name  | `browser_intake_datadoghq_com`      |
| URL   | `https://browser-intake-datadoghq.com` |

> For non-US1 Datadog sites, use the intake endpoint for the correct Datadog site.

Reviewer 2: ```suggestion
For non-US1 Datadog sites, use the intake endpoint for the correct [Datadog site][4].
```

3. Save the site.
4. Publish the site.

### 7. Validate the Theme/Layout LWC installation

After publishing the site:

1. Open the Experience Cloud site in a new browser session.
2. Open browser developer tools.
3. Confirm that the Datadog static resource loads successfully.
4. Confirm there are no CSP errors for the Datadog browser intake endpoint.
5. Navigate between site pages.
6. Confirm that RUM view events appear in Datadog.

If the component is accidentally added multiple times, the global initialization guard prevents duplicate SDK initialization, but the component should still be placed only once for clarity and maintainability.

## Salesforce Feature Support Matrix

The following table outlines SDK feature support within the Lightning Web Security (LWS) sandbox environment.

| Feature Area                | Supported    | Notes                                       |
| ---------------------------- | ------------ | -------------------------------------------- |
| **View Events**               |              |                                               |
| Initial View                 | Yes          | Automatic on init.                           |

> | Manual Tracking              | Yes          | Supported via `startView`.                   |

Reviewer 2: ```suggestion
| Manual Tracking              | Yes          | Supported through `startView`.                   |
```

| Navigation Timings           | Yes          | Collected via performance API.               |
| Web Vitals                   | Yes          |                                               |
| **Resource Events**           |              |                                               |
| Fetch / XHR                  | Limited (2)  | Context payload inaccessible.                |
| Other Resources              | Yes          | CSS, images, etc.                            |
| APM Correlation               | Limited (2)  | Requires header injection.                   |
| **Action Events**             |              |                                               |

> | Custom Actions                | Yes          | Supported via `addAction`. (5) Not supported on the Head Markup path. |

Reviewer 2: ```suggestion
| Custom Actions                | Yes          | Supported through `addAction`. (5) Not supported on the Head Markup path. |
```

| Vital Events                   | Yes          |                                               |
| Long Task Events               | Yes          |                                               |
| Session Replay                 | No           | DOM/Worker constraints prevent support.        |

Footnotes:

1. **Loading Time**: Ends when no pending network requests are detected. LWS may hide some fetch/XHR activity.
2. **Limited Context**: Inaccessible sandbox objects mean `beforeSend` cannot access response bodies or full XHR objects.
3. **Selectors**: Due to shadow boundaries, `event.target` may reflect the component host rather than the inner element.
4. **Runtime Errors**: Errors passing through the Lightning shell may lose stack traces and original error objects.
5. **Head Markup limitation**: The Experience Cloud Head Markup path has no component context to call `addAction` or `addError`, so custom actions and custom error tracking are not supported there.

## Troubleshooting

Need help? Contact [Datadog Support][3].

[1]: https://help.salesforce.com/s/articleView?id=sf.lwc_lwsec_enable.htm
[2]: https://www.datadoghq-browser-agent.com/us1/v7/datadog-rum-slim.js

> [3]: https://docs.datadoghq.com/help/

Reviewer 2: ```suggestion
[3]: https://docs.datadoghq.com/help/
[4]: https://docs.datadoghq.com/getting_started/site/#access-the-datadog-site
```
